processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -230,23 +246,31 @@ public CursorSequenceBuilder(
 
       final GenericColumn timestamps = index.getColumn(Column.TIME_COLUMN_NAME).getGenericColumn();
 
+      Iterable<Long> iterable = gran.iterable(interval.getStartMillis(), interval.getEndMillis());
+      if (descending) {
+        iterable = Lists.reverse(ImmutableList.copyOf(iterable));
+      }
       return Sequences.withBaggage(
           Sequences.map(
-              Sequences.simple(gran.iterable(interval.getStartMillis(), interval.getEndMillis())),
+              Sequences.simple(iterable),
               new Function<Long, Cursor>()
               {
                 @Override
                 public Cursor apply(final Long input)
                 {
                   final long timeStart = Math.max(interval.getStartMillis(), input);
-                  while (baseOffset.withinBounds()
-                         && timestamps.getLongSingleValueRow(baseOffset.getOffset()) < timeStart) {
+                  final long timeEnd = Math.min(interval.getEndMillis(), gran.next(input));
+                  while (baseOffset.withinBounds()) {
+                    long current = timestamps.getLongSingleValueRow(baseOffset.getOffset());
+                    if (descending ? current < timeEnd : current >= timeStart) {
```

Comment:
This change is going to add a branch inside this tight loop checking for `descending`.  It's true that branch prediction will likely do a good job with this, but we can completely eliminate this branch by doing it earlier.

Let's move the `Function` to a variable.  Then, in the conditional on lines 250:252, we can override the reference to an implementation that is specific to the descending case.